tools/makemanifest.py

```diff
@@ -0,0 +1,206 @@
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+
+from __future__ import print_function
+import sys
+import os
+import subprocess
+
+KIND_AUTO = 0
+KIND_AS_STR = 1
+KIND_AS_MPY = 2
+KIND_MPY = 3
+
+manifest_list = []
+
+class FreezeError(Exception):
+    pass
+
+def system(cmd):
+    try:
+        output = subprocess.check_output(cmd, stderr=subprocess.STDOUT)
+        return 0, output
+    except subprocess.CalledProcessError as er:
+        return -1, er.output
+
+def convert_path(path):
+    if path.startswith(':MPY:'):
+        path = TOP + path[5:]
+    return path
+
+def get_timestamp(path, default=None):
+    try:
+        stat = os.stat(path)
+        return stat.st_mtime
+    except OSError:
+        if default is None:
+            raise FreezeError('cannot stat {}'.format(path))
+        return default
+
+def get_timestamp_newest(path):
+    ts_newest = 0
+    for dirpath, dirnames, filenames in os.walk(path):
+        for f in filenames:
+            ts_newest = max(ts_newest, get_timestamp(os.path.join(dirpath, f)))
+    return ts_newest
+
+def mkdir(path):
+    cur_path = ''
+    for p in path.split('/')[:-1]:
+        cur_path += p + '/'
+        try:
+            os.mkdir(cur_path)
+        except OSError as er:
+            if er.args[0] == 17: # file exists
+                pass
+            else:
+                raise er
+
+def include(manifest):
+    if not isinstance(manifest, str):
+        for m in manifest:
+            include(m)
+    else:
+        with open(manifest) as f:
+            exec(f.read())
+
+def freeze_internal(kind, path, script, opt):
+    path = convert_path(path)
+    if script is None and kind == KIND_AS_STR:
+        if any(f[0] == KIND_AS_STR for f in manifest_list):
+            raise FreezeError('can only freeze one str directory')
+        manifest_list.append((KIND_AS_STR, path, script, opt))
+    elif script is None:
+        for dirpath, dirnames, filenames in os.walk(path):
+            for f in filenames:
+                freeze_internal(kind, path, (dirpath + '/' + f)[len(path) + 1:], opt)
+    elif not isinstance(script, str):
+        for s in script:
+            freeze_internal(kind, path, s, opt)
+    else:
+        extension_kind = {KIND_AS_MPY: '.py', KIND_MPY: '.mpy'}
+        if kind == KIND_AUTO:
+            for k, ext in extension_kind.items():
+                if script.endswith(ext):
+                    kind = k
+                    break
+            else:
+                raise FreezeError('unsupported file type {}'.format(script))
+        wanted_extension = extension_kind[kind]
+        if not script.endswith(wanted_extension):
+            raise FreezeError('expecting a {} file, got {}'.format(wanted_extension, script))
+        manifest_list.append((kind, path, script, opt))
+
+def freeze_as_str(path):
```

Comment:
Could you add a comment to these next four functions explaining what they do. (It's sort of obvious if you know the internals, but someone writing a manifest might find it useful).

Reply:
Reorded files to top and added comments on usage.